File: bench/spectral-norm.go
/* The Computer Language Benchmarks Game
 * http://benchmarksgame.alioth.debian.org/
 *
 * contributed by K P anonymous
 * corrected by Isaac Gouy
 */

package benc

import (
	"fmt"
	"math"
	"runtime"
	"time"
)

var nCPU = runtime.NumCPU() * 2

type Vec []float64

func (v Vec) Times(ii, n int, u Vec, c chan int) {
	ul := len(u)
	for i := ii; i < n; i++ {
		var vi float64
		for j := 0; j < ul; j++ {
			vi += u[j] / float64(A(i, j))
		}
		v[i] = vi
	}
	c <- 1
}

func (v Vec) TimesTransp(ii, n int, u Vec, c chan int) {
	ul := len(u)
	for i := ii; i < n; i++ {
		var vi float64
		for j := 0; j < ul; j++ {
			vi += u[j] / float64(A(j, i))
		}
		v[i] = vi
	}
	c <- 1
}

func wait(c chan int) {
	for i := 0; i < nCPU; i++ {
		<-c
	}
}

func (v Vec) ATimesTransp(u Vec) {
	x := make(Vec, len(u))
	c := make(chan int, nCPU)
	for i := 0; i < nCPU; i++ {
		go x.Times(i*len(v)/nCPU, (i+1)*len(v)/nCPU, u, c)
	}
	wait(c)
	for i := 0; i < nCPU; i++ {
		go v.TimesTransp(i*len(v)/nCPU, (i+1)*len(v)/nCPU, x, c)
	}
	wait(c)
}

func A(i, j int) int {
	return ((i+j)*(i+j+1)/2 + i + 1)
}

func SpectralNorm(n int) []string {
	start := time.Now()
	response := []string{}

	runtime.GOMAXPROCS(nCPU)

	u := make(Vec, n)
	for i := range u {
		u[i] = 1
	}
	v := make(Vec, n)
	for i := 0; i < 10; i++ {
		v.ATimesTransp(u)
		u.ATimesTransp(v)
	}
	var vBv, vv float64
	for i, vi := range v {
		vBv += u[i] * vi
		vv += vi * vi
	}
	response = append(response, fmt.Sprintf("Result: %0.9f\n", math.Sqrt(vBv/vv)))
	response = append(response, fmt.Sprintf("Execution took: %d ms\n", time.Since(start)/time.Millisecond))
	return response
}

File: bench/spectral-norm.js
// The Computer Language Benchmarks Game
// http://benchmarksgame.alioth.debian.org/
//
// contributed by Ian Osgood
// modified for Node.js by Isaac Gouy

const A = (i, j) => 1 / ((i + j) * (i + j + 1) / 2 + i + 1)

const Au = (u, v) => {
  for (let i = 0; i < u.length; ++i) {
    let t = 0
    for (let j = 0; j < u.length; ++j) {
      t += A(i, j) * u[j]
    }
    v[i] = t
  }
}

const Atu = (u, v) => {
  for (let i = 0; i < u.length; ++i) {
    let t = 0
    for (let j = 0; j < u.length; ++j) {
      t += A(j, i) * u[j]
    }
    v[i] = t
  }
}

const AtAu = (u, v, w) => {
  Au(u, w)
  Atu(w, v)
}

module.exports = (n) => {
  const start = Date.now()
  const response = []

  const u = new Float64Array(n)
  const v = new Float64Array(n)
  const w = new Float64Array(n)
  let vv = 0
  let vBv = 0
  let i

  for (i = 0; i < n; ++i) {
    u[i] = 1
    v[i] = w[i] = 0
  }

  for (i = 0; i < 10; ++i) {
    AtAu(u, v, w)
    AtAu(v, u, w)
  }

  for (i = 0; i < n; ++i) {
    vBv += u[i] * v[i]
    vv += v[i] * v[i]
  }

  response.push(`Result: ${Math.sqrt(vBv / vv).toFixed(9)}\n`)
  response.push(`Execution took: ${Date.now() - start} ms\n`)

  return response
}